gsdchem/gsd_chem_plume_wrapper.F90

```diff
@@ -59,11 +60,13 @@ subroutine gsd_chem_plume_wrapper_run(im, kte, kme, ktau, dt,                &
     integer, parameter :: ims=1,jms=1,jme=1, kms=1
     integer, parameter :: its=1,jts=1,jte=1, kts=1
 
-    logical,        intent(in) :: do_sppt_emis
-    real(kind_phys), optional, intent(in) :: sppt_wts(:,:)
+    logical,        intent(in) :: do_sppt_emis, do_ca, ca_sgs_emis, ca_sgs
+    real(kind_phys), optional, intent(in) :: sppt_wts(:,:), ca_emis_plume(:)
```

Comment:
Remove `optional`, this is the same issue I had to fix in the tendencies code. The assumed-size array declaration ensures that the code works even if `ca_emis_plume` is not allocated. An optional argument that is present and not allocated causes trouble, such as non-reproducible results (against existing code), model crashes in DEBUG mode, ... I will flag all other errors like this in the PR.

Reply:
The reason for the strange compiler-specific crashes is that CCPP's whole argument-passing paradigm violates the Fortran standard. It's pure luck that it works at all. The choice between optional arguments or no, is arbitrary, and will serve as workarounds for various compilers. I'm going to reference parts of the Fortran 2018 standard here, but these parts haven't changed appreciably since pointers and allocatables were added to the standard, so they still apply to Fortran 2008.

The first problem comes from treatment of disassociated defined (nullified) pointers. Here's section 8.5.8.4 paragraphs 1 and 3:

> 1. A deferred-shape array is an allocatable array or an array pointer. (An allocatable array has the ALLOCATABLE attribute; an array pointer has the POINTER attribute.)
> 
> R820 deferred-shape-spec  is   :
> C832 An array with the POINTER or ALLOCATABLE attribute shall have an array-spec that is a deferred-shape-spec-list.
> 
> ...
> 
> 3. The size, bounds, and shape of an unallocated allocatable array or a disassociated array pointer are undefined. No part of such an array shall be referenced or defined; however, the array may appear as an argument to an intrinsic inquiry function as specified in 16.1.

Note that the shape of the disassociated pointer array is undefined. Here's the definition of an assumed-shape array, from section 8.5.8.3, paragraph 1:

> An assumed-shape array is a nonallocatable nonpointer dummy argument array that takes its shape from its effective argument.
> 
> R819 assumed-shape-spec is [ lower-bound ] :

If you provide something with an undefined shape as an effective argument to a dummy variable, then you've violated that quite soundly. 

An explicit-shape array, of course, won't work either since it explicitly defines the shape, of a disassociated array pointer which has no shape.

Definition 3.64 states,

> 3.64 effective argument
> entity that is argument-associated with a dummy argument

In other words, if you don't pass the argument, it is not an effective argument, so the assumed-shape array has no shape.

Section 8.5.12 paragraph 1 states:

> The OPTIONAL attribute specifies that the dummy argument need not have an effective argument in a reference to the procedure (15.5.2.12).
> C849 An entity with the OPTIONAL attribute shall be a dummy argument.

Hence, optional arguments are how you make a dummy argument not be an effective argument. However, if you pass an argument at all, you have an effective argument.

The problem is that CCPP is passing array pointers without considering whether they're associated, or even defined. That means we always have an effective argument, which is a pointer (in FV3). It's being sent to a dummy argument with assumed shape, which needs an effective argument with defined shape (and cannot be a pointer or allocatable).

The only clean solution is for all arrays to be pointers. If a model wants to use non-pointers to allocate the data, they just have to make sure they use the TARGET attribute, so the array can be pointed to. The UPP solved this problem using module-scope pointers, WRF solved it using a derived type containing pointers, and ESMF solved it using pointers to nested derived types.

Reply:
I added these last-minute changes to get the code to stop crashing with non-perturbed chemistry cases. I'll see if removing the "optional" specifications still passes. It'll be interesting to see if the non-"optional" version matches the "optional" version of the code, bit-for-bit.

Reply:
> I added these last-minute changes to get the code to stop crashing with non-perturbed chemistry cases. I'll see if removing the "optional" specifications still passes. It'll be interesting to see if the non-"optional" version matches the "optional" version of the code, bit-for-bit.

It didn't for `fv3_gsd` in your last commit, and also not for verifying UGWPv1 drag suite vs GSL drag suite.